concurrency/ArchiveSerDes.java

```diff
@@ -0,0 +1,226 @@
+            resultFuture.completeExceptionally(e);
+        } finally {
+            wipExtractions.remove(destExtractionPath);
+        }
+        return resultFuture;
+    }
+
+    private void extractTgzInternal(final Path srcTgzFile, final Path destExtractionPath) {
+        try (FileInputStream fis = fileSystem.newFileInputStream(srcTgzFile.toFile());
+                GzipCompressorInputStream gzipIn = new GzipCompressorInputStream(fis);
+                TarArchiveInputStream tarIn = new TarArchiveInputStream(gzipIn)) {
+            TarArchiveEntry entry;
+            while ((entry = (TarArchiveEntry) tarIn.getNextEntry()) != null) {
+                File f = destExtractionPath.resolve(entry.getName()).toFile();
+                if (entry.isDirectory()) {
+                    mkdirs(f);
+                } else {
+                    mkdirs(f.getParentFile());
+                    try (FileOutputStream fos = fileSystem.newFileOutputStream(
+                            destExtractionPath.resolve(entry.getName()).toFile(),
+                            false)) {
+                        fileSystem.copy(tarIn, fos);
+                    }
+                }
+            }
+        } catch (IOException e) {
+            throw new ArchiveExtractionException("Failed to extract from: " + srcTgzFile, e);
+        }
+    }
+
+    private void extractZipInternal(final Path srcZipFile, final Path destExtractionPath) {
+        try (ZipFile zipFile = fileSystem.newZipFile(srcZipFile.toFile())) {
+            Enumeration<? extends ZipEntry> entries = zipFile.entries();
+            while (entries.hasMoreElements()) {
+                ZipEntry entry = entries.nextElement();
+                File entryDestination = destExtractionPath.resolve(entry.getName()).toFile();
+                if (entry.isDirectory()) {
+                    mkdirs(entryDestination);
+                } else {
+                    mkdirs(entryDestination.getParentFile());
+                    try (InputStream in = zipFile.getInputStream(entry);
+                            OutputStream out = new FileOutputStream(entryDestination)) {
+                        fileSystem.copy(in, out);
+                    }
+                }
+            }
+        } catch (IOException e) {
+            throw new ArchiveExtractionException("Failed to extract from: " + srcZipFile, e);
+        }
+    }
+
+    @VisibleForTesting
+    Path getExtractionCompletionMarkerPath(Path srcFile, Path destDir) {
+        return destDir.resolve(COMPLETION_MARKER_FILE);
+    }
+
+    private void waitForFuture(Path srcFile, Path dstDir, CompletableFuture<Void> future) {
+        try {
+            future.get(maxExtractionWaitSeconds, TimeUnit.SECONDS);
+        } catch (InterruptedException e) {
+            Thread.currentThread().interrupt();
+            throw new ArchiveExtractionException(String.format("Interrupted while extracting %s to %s", srcFile, dstDir), e);
+        } catch (ExecutionException e) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: While wrapping the caught exception into a custom one, information about the caught exception is being lost, including information about the stack trace of the exception.

**Fix**: If the caught exception object does not contain sensitive information, consider passing it as the &#34;rootCause&#34; or inner exception parameter to the constructor of the new exception before throwing the new exception. (Note that not all exception constructors support inner exceptions. Use a wrapper exception that supports inner exceptions.) 
[Learn more](https://www.ibm.com/support/pages/best-practice-catching-and-re-throwing-java-exceptions)
Similar issue at line numbers 215.